### PR TITLE
Allocated Types, Array Lengths, Array LowerBounds are global for cfa-construction

### PR DESCRIPTION
**Pull request by kbatoev**
Future improvements: dominators should calculcated for gaining right sets of Allocated Types, Array Lengths and ArrayLowerBounds (named **Sets**). Currently, when execution goes to vertex with different **Sets** new vertex is created. Exception is vertices inside cycle (Such vertices are hacked to be visited only once with unique operationalStack for cfa-construction termination).

### CODE REVIEW

File: VSharp.SILI/CFA.fs [hunk 1]
```diff
@@ -304,52 +306,66 @@ module public CFA =
...
+                    let modifiedState = {initialState with allocatedTypes = allocatedTypes; lengths = lengths; lowerBounds = lowerBounds
+                                                           currentTime = currentTime; startingTime = currentTime}
+                    let initialCilState = {cilState.Empty with state = modifiedState}
                     match cfg.offsetsDemandingCall.ContainsKey offset with
                     | true ->
                         let opCode, calledMethod = cfg.offsetsDemandingCall.[offset]
                         let callSite = {sourceMethod = cfg.methodBase; offset = offset; calledMethod = calledMethod; opCode = opCode}
                         let nextOffset, this, args, cilState' = executeSeparatedOpCode offset opCode {initialCilState with ip = Instruction offset; opStack = opStack}
-                        let dstVertex = createOrGetVertex (Instruction nextOffset, cilState'.opStack)
+                        let allocatedTypes, lengths, lowerBounds = let state = cilState'.state in state.allocatedTypes, state.lengths, state.lowerBounds
```
Comment (dvvrd) [inline VSharp.SILI/CFA.fs:346]: не очень читаемо

File: VSharp.SILI/CFA.fs [hunk 2]
```diff
@@ -304,52 +306,66 @@ module public CFA =
...
-                        bypass dstVertex
+                        bypass dstVertex stateWithArgsOnFrame.allocatedTypes stateWithArgsOnFrame.lengths stateWithArgsOnFrame.lowerBounds
                     | _ ->
                         let newStates = executeInstructions ilintptr cfg {initialCilState with ip = Instruction offset; opStack = opStack}
                         newStates |> List.iter (fun state -> currentTime <- VectorTime.max currentTime state.state.currentTime)
                         let goodStates = List.filter (fun (cilState : cilState) -> not cilState.HasException) newStates
                         let erroredStates = List.filter (fun (cilState : cilState) -> cilState.HasException) newStates
                         goodStates |> List.iter (fun (cilState' : cilState) ->
-                            let dstVertex = createOrGetVertex (cilState'.ip, cilState'.opStack)
+                            let opStack, allocatedTypes, lengths, lowerBounds = let state = cilState'.state in cilState'.opStack, state.allocatedTypes, state.lengths, state.lowerBounds
```
Comment (dvvrd) [inline VSharp.SILI/CFA.fs:359]: тоже не очень красиво оформлено

File: VSharp.SILI/CFA.fs [hunk 3]
```diff
@@ -425,18 +441,17 @@ type StepInterpreter() =
...
         dfs 0u cfa.body.entryPoint
-        let resultStates = List.init (maxBorder |> int) (fun lvl -> cfa.body.exitVertex.Paths.OfLevel true (lvl |> uint32) |> List.ofSeq)
+        let resultStates = List.init (maxBorder |> int) (fun lvl -> cfa.body.exitVertex.Paths.OfLevel false (lvl |> uint32) |> List.ofSeq)
                          |> List.concat
                          |> List.map (fun (path : path) ->
                              let state = path.state
                              match state.returnRegister with
                              | None -> Nop, state
                              | Some res -> res, state)
+        if List.length resultStates = 0 then internalfailf "Most likely such a situation is a bug. Check it!"
```
Comment (dvvrd) [inline VSharp.SILI/CFA.fs:451]: Добавь в начало "No states were obtained."

File: VSharp.Test/Tests/PDR.cs [hunk 4]
```diff
@@ -221,11 +222,170 @@ private static int F(int x)
         }
 
         [TestSvm]
-        public static int FunctionsComposition(int x)
+        public static int SolverTestConcreteArray(int x)
```
Comment (dvvrd) [inline VSharp.Test/Tests/PDR.cs:225]: Вопрос скорее к Мише, но при чём тут солвер?

File: VSharp.SILI/CFA.fs [hunk 5]
```diff
@@ -425,18 +445,17 @@ type StepInterpreter() =
         cfa.body.entryPoint.Paths.Add {lvl = 0u; state = initialState}
         Logger.trace "starting Forward exploration for %O" codeLoc
         dfs 0u cfa.body.entryPoint
-        let resultStates = List.init (maxBorder |> int) (fun lvl -> cfa.body.exitVertex.Paths.OfLevel true (lvl |> uint32) |> List.ofSeq)
+        let resultStates = List.init (maxBorder |> int) (fun lvl -> cfa.body.exitVertex.Paths.OfLevel false (lvl |> uint32) |> List.ofSeq)
```
Comment (MchKosticyn) [inline VSharp.SILI/CFA.fs:448]: Need to pass shouldGetAllPaths variable here?

File: VSharp.Utils/MappedStack.fs [hunk 6]
```diff
@@ -35,15 +35,23 @@ module public MappedStack =
...
     let remove (contents, peaks) key =
         let idx = peakIdx peaks key
-        assert (idx > defaultPeak)
-        let key' = makeExtendedKey key idx
-        let contents' = Map.remove key' contents
-        let peaks' =
-            if idx = 1ul then Map.remove key peaks
-            else Map.add key (idx - 1ul) peaks
-        contents', peaks'
+        match containsKey key (contents, peaks) with
```
Comment (MchKosticyn) [inline VSharp.Utils/MappedStack.fs:42]: Need to use if instead of match

**Review by MchKosticyn — APPROVED** [review]

**Pull request merged**